scrape_books/middlewares.py

```diff
@@ -0,0 +1,70 @@
+from typing import Iterable
+
+from scrapy import signals, Spider, Request
+from scrapy.crawler import Crawler
+from scrapy.http import Response
+
+
+class ScrapeBooksSpiderMiddleware:
+
+    @classmethod
+    def from_crawler(cls, crawler: Crawler) -> Spider:
+        spider = cls()
+        crawler.signals.connect(
+            spider.spider_opened, signal=signals.spider_opened
+        )
+        return spider
+
+    def process_spider_output(
+            self,
+            response: Response,
+            result: Iterable,
+            spider: Spider
+    ) -> None:
+        for i in result:
+            yield i
+
+    def process_spider_exception(
+            self,
+            response: Response,
+            exception: Exception,
+            spider: Spider
+    ) -> None:
+        spider.logger.error(
+            f"Exception encountered: {exception} on {response.url}"
+        )
+        return None
```

Comment:
The `process_spider_exception` method logs exceptions but does not handle them further. If specific handling is required, consider implementing additional logic here.